### ISSUE TITLE
java.nio.channels.CancelledKeyException

### ISSUE THREAD

**Issue by RoyJing** [dnsjava/dnsjava#129]
After our app was released, we had a lot of crashes like this. The SDK version is 3.0.2, this detailed stack information

Fatal Exception: java.nio.channels.CancelledKeyException
       at sun.nio.ch.SelectionKeyImpl.ensureValid(SelectionKeyImpl.java:73)
       at sun.nio.ch.SelectionKeyImpl.readyOps(SelectionKeyImpl.java:87)
       at java.nio.channels.SelectionKey.isReadable(SelectionKey.java:289)
       at org.xbill.DNS.NioUdpClient$Transaction.processReadyKey(NioUdpClient.java:119)
       at org.xbill.DNS.Client.processReadyKeys(Client.java:79)
       at org.xbill.DNS.Client.runSelector(Client.java:62)
       at org.xbill.DNS.Client.lambda$jAWiqsJXxSKrE-3ZAhrF3bfbwqQ(Client.java)
       at o.-$$Lambda$fnu$jAWiqsJXxSKrE-3ZAhrF3bfbwqQ.run(-.java)
       at java.lang.Thread.run(Thread.java:764)

**Comment by ibauersachs** [comment]
The current version is 3.2.2. I cannot support older versions and the particular issue was AFAIK fixed in 3.1.0.

**Comment by RoyJing** [comment]
> The current version is 3.2.2. I cannot support older versions and the particular issue was AFAIK fixed in 3.1.0.

Great, I'll upgrade to the latest version

**Comment by ibauersachs** [comment]
@AlexanderGH Do you have more information when this happens? I found an old JDK bug (https://bugs.openjdk.org/browse/JDK-4729342) which says that "old" selection keys should be cancelled. We don't do this from what I can tell at a brief glance.
It would help tremendously if you could come up with a reproducer, even if it only happens in 1 out of 10 runs.

**Comment by AlexanderGH** [comment]
So sadly this only happens in an Android app in prod. As best we can tell, this correlates to when we System.exit(0) or possibly restart the app in the background after System.exit(0)ing.

Stats:
- It happens on Android versions all the way up to 16, so if it is a bug in ART, it hasn't been fixed.
- Doesn't look like it's caused by an OEM modification since it happens on many OEMs

**Comment by ibauersachs** [comment]
The exit part at least correlates with what I'd assume could happen as a race condition. Initially I thought this was during normal lookups while the application was running.

Can you test if the following patch helps?

```patch
diff --git a/src/main/java/org/xbill/DNS/NioUdpClient.java b/src/main/java/org/xbill/DNS/NioUdpClient.java
--- a/src/main/java/org/xbill/DNS/NioUdpClient.java	(revision b9180e461d826ef526df7f968ce689cefe4701a4)
+++ b/src/main/java/org/xbill/DNS/NioUdpClient.java	(date 1768759896467)
@@ -109,9 +109,14 @@
 
     @Override
     public void processReadyKey(SelectionKey key) {
+      if (!key.isValid()) {
+        return;
+      }
+
       if (!key.isReadable()) {
         completeExceptionally(new EOFException("Key for transaction " + id + " is not readable"));
         pendingTransactions.remove(this);
+        key.cancel();
         return;
       }
 
@@ -121,11 +126,15 @@
       try {
         read = keyChannel.read(buffer);
         if (read <= 0) {
-          throw new EOFException();
+          completeExceptionally(
+              new EOFException("Could not read expected data for transaction " + id));
+          pendingTransactions.remove(this);
+          key.cancel();
         }
       } catch (IOException e) {
         completeExceptionally(e);
         pendingTransactions.remove(this);
+        key.cancel();
         return;
       }
 
@@ -137,6 +146,7 @@
           keyChannel.socket().getLocalSocketAddress(),
           keyChannel.socket().getRemoteSocketAddress(),
           resultingData);
+      key.cancel();
       silentDisconnectAndCloseChannel();
       f.complete(resultingData);
       pendingTransactions.remove(this);
```

**Comment by ibauersachs** [comment]
If it only happens during exit(), there's a good chance that 3.6.4 solves it without the above (untested) patch.

Yes, calling close() is the same as the shutdown hook. You'd need to make sure to call close from a different thread though (like in the shutdown hook).

**Comment by sidakpreetnanda** [comment]
Hey @ibauersachs 

Can you please release the fix for this? v3.6.4 does not seem to have this
https://github.com/dnsjava/dnsjava/commit/c1478fa58f0e863d410d9c8f47122b1825722e40
We are facing crashes due to same in production.

`Fatal Exception: java.nio.channels.CancelledKeyException:
       at sun.nio.ch.SelectionKeyImpl.ensureValid(SelectionKeyImpl.java:73)
       at sun.nio.ch.SelectionKeyImpl.readyOps(SelectionKeyImpl.java:87)
       at java.nio.channels.SelectionKey.isReadable(SelectionKey.java:289)
       at org.xbill.DNS.NioUdpClient$Transaction.processReadyKey(NioUdpClient.java:112)
       at org.xbill.DNS.NioClient.processReadyKeys(NioClient.java:182)
       at org.xbill.DNS.NioClient.runSelector(NioClient.java:134)
       at java.lang.Thread.run(Thread.java:919)`

**Comment by AlexanderGH** [comment]
Still seeing this sadly with 3.6.5:

```
java.nio.channels.CancelledKeyException
        at sun.nio.ch.SelectionKeyImpl.ensureValid(SelectionKeyImpl.java:73)
        at sun.nio.ch.SelectionKeyImpl.interestOps(SelectionKeyImpl.java:82)
        at org.xbill.DNS.NioTcpClient$ChannelState.processWrite(NioTcpClient:282)
        at org.xbill.DNS.NioTcpClient$ChannelState.processReadyKey(NioTcpClient:148)
        at org.xbill.DNS.NioClient.processReadyKeys(NioClient:255)
        at org.xbill.DNS.NioClient.runSelector(NioClient:156)
        at org.xbill.DNS.NioClient$$ExternalSyntheticLambda0.run(NioClient:100)
        at java.lang.Thread.run(Thread.java:1564)
```
